ssm_jax/hmm/core.py

```diff
@@ -0,0 +1,226 @@
+import jax.numpy as jnp
+import jax.random as jr
+from jax import lax
+
+
+def hmm_filter(initial_distribution,
+               transition_matrix,
+               log_likelihoods):
+    """_summary_
+
+    Args:
+        hmm_params (_type_): _description_
+        log_likelihoods (_type_): _description_
+
+    Returns:
+        _type_: _description_
+    """
+    def _step(carry, t):
+        log_normalizer, pred_probs = carry
+
+        # Get parameters for time t
+        get = lambda x, ndim: x[t] if x.ndim == ndim + 1 else x
+        A = get(transition_matrix, 2)
```

Comment:
I would suggest
```
get_params = lambda A, t: A[t] if A.ndim == 3 else A 
A = get_params(transition_matrix, t)
```
Or
```
get_params = lambda A, t: A(t) if iscallable(A) else A 
A = get_params(transition_matrix, t)
```

Reply:
I was thinking about something like this. One concern is that JAX won't let you jit a function that takes a function as input ("not a valid JAX type"), so I thought for these low-level interfaces I'd assume the inputs were always arrays. This entails a bit of memory inefficiency for nonstationary cases where you have to construct all the transition matrices in memory, but it's a simpler interface.

Reply:
Makes sense.